### PR TITLE
ENT-12140: deps upgrades (3.21)

### PR DESCRIPTION
**Pull request by craigcomstock**
Ticket: ENT-12140

together
https://github.com/cfengine/core/pull/5603
https://github.com/cfengine/buildscripts/pull/1491

### CODE REVIEW

**Comment by craigcomstock** [comment]
[Build Status](https://ci.cfengine.com/job/pr-pipeline/11137/)

**Comment by craigcomstock** [comment]
@cf-bottom jenkins please, with fixes for centos-7 and ubuntu-20

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/11138/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/11138/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-11138/

**Comment by craigcomstock** [comment]
Try just centos-7 with fixes: [Build Status](https://ci.cfengine.com/job/pr-pipeline/11139/)

**Comment by craigcomstock** [comment]
[Build Status](https://ci.cfengine.com/job/pr-pipeline/11140/)

**Comment by craigcomstock** [comment]
[Build Status](https://ci.cfengine.com/job/pr-pipeline/11141/)

**Comment by craigcomstock** [comment]
[Build Status](https://ci.cfengine.com/job/pr-pipeline/11142/)

**Comment by craigcomstock** [comment]
[Build Status](https://ci.cfengine.com/job/pr-pipeline/11143/)

**Comment by craigcomstock** [comment]
[Build Status](https://ci.cfengine.com/job/pr-pipeline/11144/)

**Comment by craigcomstock** [comment]
[Build Status](https://ci.cfengine.com/job/pr-pipeline/11145/)

**Comment by craigcomstock** [comment]
[Build Status](https://ci.cfengine.com/job/pr-pipeline/11146/)

**Comment by craigcomstock** [comment]
[Build Status](https://ci.cfengine.com/job/pr-pipeline/11148/)

**Comment by craigcomstock** [comment]
[Build Status](https://ci.cfengine.com/job/pr-pipeline/11151/)

**Comment by craigcomstock** [comment]
added more things to package in the test-only package: [Build Status](https://ci.cfengine.com/job/pr-pipeline/11152/)

**Comment by craigcomstock** [comment]
maybe done? with lcov at least?
[Build Status](https://ci.cfengine.com/job/pr-pipeline/11153/)

**Comment by craigcomstock** [comment]
retry, remove more things: [Build Status](https://ci.cfengine.com/job/pr-pipeline/)

**Comment by craigcomstock** [comment]
retry with lcov removal [Build Status](https://ci.cfengine.com/job/pr-pipeline/11157/)

**Comment by craigcomstock** [comment]
rebuilding for rhel8hub: [Build Status](https://ci.cfengine.com/job/pr-pipeline/11158/)

**Comment by craigcomstock** [comment]
Trying a full build as maybe things are good now?
[Build Status](https://ci.cfengine.com/job/pr-pipeline/11160/)

**Comment by craigcomstock** [comment]
rhel8hub only: [Build Status](https://ci.cfengine.com/job/pr-pipeline/11162/)

**Comment by craigcomstock** [comment]
retry ubu16hub: [Build Status](https://ci.cfengine.com/job/pr-pipeline/11163/)

**Comment by craigcomstock** [comment]
oops, forgot the mission-portal PR in the rhel8hub rebuild, should be fine, but will re-check now: 
[Build Status](https://ci.cfengine.com/job/pr-pipeline/11164/)

**Comment by craigcomstock** [comment]
yeah, looooooooks good. @cf-bottom jenkins please, throw everything at this PR. Thanks.

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/11177/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/11177/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-11177/

File: deps-packaging/php/cfbuild-php.spec [hunk 1]
```diff
@@ -34,7 +34,7 @@ fi
   --with-iconv \
   --with-zlib=%{prefix} \
   --with-libmbfl=%{prefix} \
-  --enable-mbstring \
+  --enable-mbstring=shared \
```
Comment (vpodzime) [inline deps-packaging/php/cfbuild-php.spec:37]: Are we also doing:
> you also need to add extension=mbstring.so to your php.ini
?

**Review by vpodzime — APPROVED** [review]
Looks good to me.

File: deps-packaging/php/cfbuild-php.spec (hunk 1 above)
Comment (craigcomstock) [reply deps-packaging/php/cfbuild-php.spec:37]: no. curious about where/how it is used! :+1: I will check. I didn't see any issues/errors...

Comment (craigcomstock) [reply deps-packaging/php/cfbuild-php.spec:37]: we have three shared extensions:
```
cfbuild-php.spec:  --with-openssl=shared,%{prefix} \
cfbuild-php.spec:  --with-curl=shared,%{prefix} \
cfbuild-php.spec:  --enable-mbstring=shared \
```

Interesting that the others had a trailing `,%{prefix}` and neither curl nor openssl .so files are explicitly added to php.ini.

I will check that the `mb_*` functions operate properly on an instance and see if we need to fix things. :+1: The `mb_*` functions are only used in a handful of places but a few may be used by many other places.

```
application/libraries/Cfe_table.php:                                        $data = urlencode((mb_convert_encoding($data, 'UTF-8')));
application/controllers/Hubstatus.php:        $arrayData = json_decode(mb_convert_encoding($statusData, 'UTF-8'), true);
tests/mocks/ci_basetestcase.php:        $array = json_decode(mb_convert_encoding($data, 'UTF-8'), true);
tests/mocks/core/utf8.php:                      mb_internal_encoding('UTF-8');
tests/phpunit/models/VitalsModelTest.php:        $array = json_decode(mb_convert_encoding($json, 'UTF-8'), true);
```

**Review by craigcomstock — COMMENTED** [review]
will check shared extensions functioning

File: deps-packaging/php/cfbuild-php.spec (hunk 1 above)
Comment (craigcomstock) [reply deps-packaging/php/cfbuild-php.spec:37]: yeah, good catch, this sample doesn't work unless I add `extension=mbstring.so` to a local `php.ini` file.

I wonder about the other two: curl and openssl, will check!

Ah, we handle this in a different way: individual files:

```
root@localhost:~# php --ini
Configuration File (php.ini) Path: /var/cfengine/httpd/php/lib
Loaded Configuration File:         /var/cfengine/httpd/php/lib/php.ini
Scan for additional .ini files in: /var/cfengine/httpd/php/lib
Additional .ini files parsed:      /var/cfengine/httpd/php/lib/cfengine-enterprise-api.ini,
/var/cfengine/httpd/php/lib/cfmod.ini,
/var/cfengine/httpd/php/lib/curl.ini,
/var/cfengine/httpd/php/lib/openssl.ini,
/var/cfengine/httpd/php/lib/php.ini
```

Will fix.

Comment (craigcomstock) [reply deps-packaging/php/cfbuild-php.spec:37]: https://www.phpinternalsbook.com/php7/build_system/building_extensions.html#building-php-extensions

not sure what the `,%{prefix}` is all about yet, looking...

Comment (craigcomstock) [reply deps-packaging/php/cfbuild-php.spec:37]: from `configure` script output

```

Extensions:

  --with-EXTENSION=shared[,PATH]

    NOTE: Not all extensions can be built as 'shared'.

    Example: --with-foobar=shared,/usr/local/foobar/

      o Builds the foobar extension as shared extension.
      o foobar package install prefix is /usr/local/foobar/
```

So given the result of the latest build the PREFIX added on doesn't do much as all three shared extensions are in the same place. I'll add it as I fix the inclusion now and see how it goes.

**Comment by craigcomstock** [comment]
cancelled the big CI run and retrying ubu16hub only: [Build Status](https://ci.cfengine.com/job/pr-pipeline/11178/)

**Comment by craigcomstock** [comment]
@cf-bottom jenkins again please, full deal, I think we are good. :crossed_fingers:

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/11179/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/11179/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-11179/

**Comment by craigcomstock** [comment]
retry build tests, seem flaky: [Build Status](https://ci.cfengine.com/job/sequential-tests/7617/)

**Comment by craigcomstock** [comment]
check aix71 since it was offline
[Build Status](https://ci.cfengine.com/job/pr-pipeline/11214/)

**Pull request merged**